### DISCUSSION TITLE
Getting unverified signature even if git says the signature is ok, uploaded and with the same email and I used to get verified

### DISCUSSION THREAD

**Discussion by gurghet** [community/community#59941] [Code Security] [answered]
### Select Topic Area

Bug

### Body

Exactly as the title says, I used to get verified for my commits. Now I sign every commit, I get unverified. I reuploaded the key, which checks out locally, and commits are good and verified. Email corresponds, and also checked that the key is the same.
I don't know why it keeps showing unverified, how to properly troubleshoot this? How do I compare what I have locally with what Github receives?

**Comment by 3urobeat** [comment]
Hey, I took a look at one of your commits and GitHub does not list a GPG key attached to the commit. Since you have vigilant mode enabled all unsigned commits are marked as `Unsigned`.  

Make sure that in your `.gitconfig` `commit.gpgsign` is `true` and that your correct `signingkey` is set under `user`.  
This [GitHub docs article](https://docs.github.com/en/authentication/managing-commit-signature-verification/telling-git-about-your-signing-key#telling-git-about-your-gpg-key) explains how you can set your key and enable signing.  

You can also take a look at if git signed your commit locally:  
Open a terminal in your git repository folder and run the command `git show --format=fuller --show-signature <sha-of-a-commit-from-this-repo>`.  
It should list a gpg key at the top of the output if the commit was signed.

I hope this helps you narrowing down your issue.

**Reply by gurghet** [reply]
Thanks, it does, the .gitconfig is correct, and the commit shows a good signature with the corresponding GPG key, which is attached to the commit. How do you list GPG keys in github commits?

**Reply by 3urobeat** [reply]
I think the only way to check the GPG keys in your GitHub commit is by clicking on the `Unverified` icon in your commit list.  
It either lists a reason why your commit appears as unverified or says that it wasn't signed.  

Have you taken a look at your [GitHub GPG settings](https://github.com/settings/keys)? Does it maybe say `Unverified` behind the email associated to your commit?  
Have you recently changed your GitHub username?  
Are you using the same Email in your `.gitconfig` `user` settings as the GPG key uses? And is this email added to your GitHub account in the [email settings](https://github.com/settings/emails)?  

There was also once a "bug" with GitHub Desktop and signing, should you use that software to make your commits. I'm not sure if that is fixed already (but I think so).

That's a few more reasons I could think of right now why your commit may be unverified.

**Reply by gurghet** [reply]
I click on unverified and says ```The signature in this commit could not be verified. Someone may be trying to trick you.
GPG key ID: <key-id-A>
[Learn about vigilant mode](https://docs.github.com/github/authenticating-to-github/displaying-verification-statuses-for-all-of-your-commits).```
Then I look into my settings and it says:
```
GPG
General signatory key
Email address:  gurghet@gmail.com
Key ID: <key-id-A>
Subkeys: <key-id-B>
Added on some past date
```
then locally I have
```
[user]
	name = Andrea Passaglia
	email = gurghet@gmail.com
	signingkey = <key-id-A>
...
[gpg]
	program = /Users/gurghet/Developer/bin/gpg-with-pass.sh
```
said program contains
```
#!/bin/zsh
export GPG_TTY=$(tty)
passphrase=$(op item get 'GPG Passphrase' --format json | jq '.fields[0].value' -r)
echo $passphrase | gpg --pinentry-mode loopback --passphrase-fd 0 "$@"
```

**Reply by 3urobeat** [reply]
You could try using your GitHub noreply email address instead of your personal one and see if it makes a difference.  
Add it to your gpg key as secondary uuid, it should be [explained here pretty well](https://docs.github.com/en/authentication/managing-commit-signature-verification/associating-an-email-with-your-gpg-key) and set it as user email in your `.gitconfig`.
It is listed in your GitHub email settings when you have "Keep my email addresses private" enabled.  

If that still doesn't work I'm kind of out of ideas, sorry.

**Reply by gurghet** [reply]
Thanks a lot, turns out it was the script, for some reason is working for local signature verfication but sends something weird to GitHub evidently

**Reply by 3urobeat** [reply]
Oh well, but good to hear that you figured it out, I hope I was at least somewhat helpful ^^

**Reply by gurghet** [reply]
of course thanks a lot! If I understand what is the diff i'll post it here

**Comment by gurghet** [comment, marked as answer]
Got the solution: I was passing *both* passphrase and commit on fd 0 (stdin) this is the solution
`exec 3<<<"$passphrase"; gpg --pinentry-mode loopback --passphrase-fd 3 "$@"`
